compiler/rustc_typeck/src/check/fn_ctxt/checks.rs

```diff
@@ -1226,7 +1226,22 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     }
 
     /// Type check a `let` statement.
-    pub fn check_decl_local(&self, local: &'tcx hir::Local<'tcx>) {
+    pub fn check_decl_local(
+        &self,
+        local: &'tcx hir::Local<'tcx>,
+        els: Option<&'tcx hir::Block<'tcx>>,
+    ) {
+        if let Some(blk) = els {
+            let previous_diverges = self.diverges.get();
+            let else_ty = self.check_block_with_expected(blk, NoExpectation);
+            let cause = self.cause(blk.span, ObligationCauseCode::LetElse);
+            if let Some(mut err) =
+                self.demand_eqtype_with_origin(&cause, self.tcx.types.never, else_ty)
+            {
+                err.emit();
+            }
+            self.diverges.set(previous_diverges);
+        }
         self.check_decl(local.into());
```

Comment:
why is the `else` block checked before the initializer?

Reply:
I did not pay attention to. It is now moved below.